api/src/main/java/org/hawkular/inventory/api/model/Path.java

```diff
@@ -0,0 +1,980 @@
+     *
+     * @return this instance converted to RelativePath
+     */
+    public abstract RelativePath toRelativePath();
+
+    /**
+     * Tries to convert this path to a canonical path. This can result in a new instance construction if this instance
+     * is a relative path.
+     *
+     * @return this instance converted to CanonicalPath
+     * @throws IllegalArgumentException if this instance is a relative path and it cannot be converted to canonical path
+     */
+    public abstract CanonicalPath toCanonicalPath();
+
+    /**
+     * @return true if this is an instance of {@link CanonicalPath}, false otherwise
+     */
+    public boolean isCanonical() {
+        return this instanceof CanonicalPath;
+    }
+
+    /**
+     * @return true if this is an instance of {@link RelativePath}, false otherwise
+     */
+    public boolean isRelative() {
+        return this instanceof RelativePath;
+    }
+
+    /**
+     * Checks whether the path is well-formed. Going {@link #up()} or {@link #down()} on the path can render it
+     * undefined if it is traversed "too far" in either of the directions.
+     *
+     * @return true if this path is well-formed, false otherwise.
+     */
+    public boolean isDefined() {
+        return endIdx > startIdx && endIdx <= path.size();
+    }
+
+    /**
+     * Returns a path corresponding to the direct ancestor of the resource represented by this path object.
+     *
+     * @return the ancestor path (may be {@link #isDefined() undefined}.
+     */
+    public Path up() {
+        return up(1);
+    }
+
+    /**
+     * Returns a path corresponding to the n-th ancestor in the hierarchy.
+     *
+     * @param distance the distance of the ancestor from the resource represented by this path object.
+     * @return the ancestor path (may be {@link #isDefined() undefined}.
+     */
+    public Path up(int distance) {
+        return newInstance(startIdx, endIdx - distance, path);
+    }
+
+    /**
+     * If this path was created by going {@link #up() up} from another path, then this method can be used to go back
+     * down to the previous paths representing some child resource of the resource represented by this path object.
+     *
+     * @return a path to a direct child of the resource represented by this path (may be {@link #isDefined() undefined}.
+     */
+    public Path down() {
+        return down(1);
+    }
+
+    /**
+     * If this path was created by going {@link #up() up} from another path, then this method can be used to go back
+     * down n steps to the previous paths representing some (grand-)child resource of the resource represented by
+     * this path object.
+     *
+     * @param distance the distance from this path to the child path
+     * @return a path to a child of the resource represented by this path (may be {@link #isDefined() undefined}.
+     */
+    public Path down(int distance) {
+        return newInstance(startIdx, endIdx + distance, path);
```

Comment:
Yes, the returned path will be undefined. Doing `path.down().up()` should return a path equivalent to the original.